shape.c

```diff
@@ -196,6 +209,7 @@ rb_shape_transition_shape_frozen(VALUE obj)
 rb_shape_t *
 rb_shape_get_next_iv_shape(rb_shape_t* shape, ID id)
 {
+    RUBY_ASSERT(is_junk_id(id) || RTEST(rb_sym2str(ID2SYM(id))));
```

Comment:
`id_debug_created_info` looks like one of those special ids that is < `tLAST_OP_ID`, if that helps at all. `is_notop_id` checks for those.

Reply:
It's just some value in an enum.  There is no associated string for that particular ID.  Not all IDs have string representations, so we need to dump something else as the edge name.

It's defined in `id.h`.  Unfortunately that file is generated, but you can find the source for it [here](https://github.com/tenderlove/ruby/blob/gc-compact/template/id.h.tmpl#L65).

When we derive the key name, we should probably look at the shape type.  Only IVAR shapes will have an edge name that _may_ have a string representation, but it's not guaranteed.  As @composerinteralia, I think `is_notop_id` is the right check.

Reply:
We looked through `id.def` and `id.c.tmpl` and it looks like `id_debug_created_info` is the only id that works like this 😓. It's defined in `id.def` to be `-`, which carries the special meaning that it has [no defined string](https://github.com/ruby/ruby/blob/master/defs/id.def#L175-L178) which [looks like it was done to hide them from marshal](https://github.com/ruby/ruby/commit/6964c0635f76ea79a49c8dcb85f279e0b85fc0cd), which skips them by [just checking if there is a string associated](https://github.com/ruby/ruby/blob/master/marshal.c#L599).

The easiest path forward is probably to duplicate that check and skip any id where `!rb_id2str(id);`

In the future we might consider having a different way for marshal to skip undesirable ivars (I'd be interested in looking into this, but not so close to release 😅).

Reply:
Actually, if we wanted to be conservative we could print out only ids where `is_instance_id(x)` (only the ones which start with an `@`), we might miss some internal things, but that should be really safe to do (and since this is for debugging we may not need it perfect)

Reply:
Ok, I tried  a few more things, but ultimately ended up using `is_instance_id` like you suggested. Thanks for the help!